src/main/InternalProcess.js

```diff
@@ -1,120 +1,212 @@
+  configure(path: string, args: ?(string[]) = [], options: ?ForkOptions = {}) {
     this.path = path;
     this.args = args;
     this.options = options;
   }
 
-  start() {
-    if (this.process) {
-      throw new Error("Internal process is already running !");
+  setPassword(password: string) {
+    if (password !== this.password) {
+      this.password = password;
+      logger.info("Libcore password set", logType);
+
+      if (this.process) {
+        return this.restart();
+      }
     }
+  }
 
-    this.process = fork(this.path, this.args, this.options);
+  changePassword(newPassword: string) {
+    logger.info("Changing Libcore password", logType);
 
-    this.active = true;
-    const pid = this.process.pid;
+    if (!this.libcoreInitialized) {
+      throw new Error("Can't change password if libcore is not unlocked first!");
+    }
 
-    logger.info(`spawned internal process ${pid}`);
-    console.log(`spawned internal process ${pid}`);
+    this.process &&
+      this.process.send({
+        type: "changeLibcorePassword",
+        currentPassword: this.password,
+        newPassword,
+      });
 
-    this.process.on("exit", (code, signal) => {
-      this.process = null;
+    this.password = newPassword;
+  }
 
-      if (code !== null) {
-        console.log(`internal process ${pid} gracefully exited with code ${code}`);
-        logger.info(`Internal process ${pid} ended with code ${code}`);
-      } else {
-        console.log(`internal process ${pid} got killed by signal ${signal}`);
-        logger.info(`Internal process ${pid} killed with signal ${signal}`);
-      }
+  start() {
+    logger.info("Started internal process", logType);
 
-      if (this.onExitCallback) {
-        this.onExitCallback(code, signal, this.active);
+    return new Promise<any>((resolve, reject) => {
+      if (this.process) {
+        throw new Error("Internal process is already running !");
       }
-    });
 
-    this.process.on("message", message => {
-      if (this.onMessageCallback && this.active) {
-        this.onMessageCallback(message);
+      if (!this.path || !this.args || !this.options) {
+        throw new Error("configure() was not completed before start()");
       }
-    });
 
-    if (this.messageQueue.length) {
-      this.run();
-    }
+      this.process = fork(this.path, this.args, this.options);
+
+      this.startPromise = { resolve, reject };
+
+      this.active = true;
+      const pid = this.process.pid;
 
-    this.process.stdout.on("data", data =>
-      String(data)
-        .split("\n")
-        .forEach(msg => {
-          if (!msg) return;
-          if (process.env.INTERNAL_LOGS) console.log(msg);
-          try {
-            const obj = JSON.parse(msg);
-            if (obj && obj.type === "log") {
-              logger.onLog(obj.log);
-              return;
+      logger.info(`spawned internal process ${pid}`, logType);
+
+      this.process && // A bit stupid, but Flow complains otherwise
```

Comment:
would it help to make `process` a tmp variable of this scope so you don't need to check this? I think this happens because flow can't prove you didn't re-set this.process in another part of the code but if you do a local `const` it will not make you do this.

this would help on reducing the diff of this PR here.

Reply:
Makes sense 👍